core/src/validator.rs

```diff
@@ -1423,16 +1433,8 @@ fn load_blockstore(
         ledger_signal_receiver,
         completed_slots_receiver,
         ..
-    } = Blockstore::open_with_signal(
-        ledger_path,
-        BlockstoreOptions {
-            recovery_mode: config.wal_recovery_mode.clone(),
-            column_options: config.ledger_column_options.clone(),
-            enforce_ulimit_nofile: config.enforce_ulimit_nofile,
-            ..BlockstoreOptions::default()
-        },
-    )
-    .expect("Failed to open ledger database");
+    } = Blockstore::open_with_signal(ledger_path, blockstore_options_from_config(config))
```

Comment:
I think we should keep the logic external to `Blockstore`; having a `Blockstore` function that takes in `ValidatorConfig` would mean that `solana_ledger` crate depends on `solana_validator` crate, which would create a circular dependency.

Reply:
If that's the case, should we have a dedicated crate for Validator configs / options where all other crates depend on?
I feel if we have an API that doesn't honor the config, then it means we have one more spot that may cause problems.

Reply:
So validator and ledger-tool are really the only files that open a `Blockstore` aside from test code. Validator has a different set of config options than ledger-tool, so any config couldn't be shared here.

I'd say the function that I've implemented in this PR is basically doing what you're talking about; it takes a Validator config and yields a BlockstoreOptions. I think it is the responsibility of Validator to do the conversion of its' config to Blockstore config

Reply:
But this leaves the possibility to invoke `Blockstore::open_with_signal` without using `blockstore_options_from_config`.  Just want to rule out all possible ways that could go wrong.

Reply:
> But this leaves the possibility to invoke Blockstore::open_with_signal without using blockstore_options_from_config

Technically yes, but even if we write some `Blockstore::open*` variations that take a `ValidatorConfig` in, we still have the generic `Blockstore` API for other use cases (ledger-tool and tests, and maybe others that I'm not thinking of).

If we add some `Blockstore::open_for_validator*()`, we just shift the problem from:
- "user may not call `blockstore_options_from_config()` in `validator.rs`"

to 
- "user may call `Blockstore::open*()` instead of `Blockstore::open_for_validator*()`

This doesn't seem any better to me in addressing your concern given that the user still has to "choose correctly". Arguably, we could adjust `Blockstore` API to only have functions like `open_from_validator()`, `open_from_ledger_tool`, ... that are catered for use cases, but as mentioned, this couples crates together and means we have to adjust the API (or abuse the API) when we come up with new cases. If you have a good proposal in this regard, I'm open to hearing it, but I'm not sure this PR is the place as I believe to do it properly would be more widespread than this very targeted bug fix.

Reply:
Oh oh I meant it isn't a block for this PR as the PR already fix one existing problem, and that's why I've approved this PR.

I am just using this opportunity to discuss whether we can have a better API.

Reply:
Ahhh, got it. Thanks for clarifying. I'll ship this one and we can mull on a broader change to prevent this type of bug from being able to occur.